File: frontend/src/services/reservationService.ts
import api from '@/lib/api'

export interface CreateReservationRequest {
  eventId: number
  userId: string // Changé en string pour matcher Keycloak
}

export interface ReservationResponse {
  id: number
  eventId: number
  userId: string
  createdAt: string // Match le backend Java
  status: 'PENDING' | 'CONFIRMED' | 'CANCELLED' | string
  seatNumber?: number | null
}

const RESERVATIONS = '/api/reservations'

export const reservationService = {
  createReservation: async (request: CreateReservationRequest): Promise<ReservationResponse> => {
    const response = await api.post<ReservationResponse>(RESERVATIONS, request)
    return response.data
  },

  getAllReservations: async (): Promise<ReservationResponse[]> => {
    const response = await api.get<ReservationResponse[]>(RESERVATIONS)
    return response.data
  },

  getReservationsByUser: async (userId: string): Promise<ReservationResponse[]> => {
    const response = await api.get<ReservationResponse[]>(`${RESERVATIONS}/user/${userId}`)
    return response.data
  },

  getReservationsByEvent: async (eventId: number): Promise<ReservationResponse[]> => {
    const response = await api.get<ReservationResponse[]>(`${RESERVATIONS}/event/${eventId}`)
    return response.data
  },

  getReservationById: async (id: number): Promise<ReservationResponse> => {
    const response = await api.get<ReservationResponse>(`${RESERVATIONS}/${id}`)
    return response.data
  },

  confirmReservation: async (id: number): Promise<ReservationResponse> => {
    const response = await api.put<ReservationResponse>(`${RESERVATIONS}/${id}/confirm`)
    return response.data
  },

  cancelReservation: async (id: number): Promise<ReservationResponse> => {
    const response = await api.put<ReservationResponse>(`${RESERVATIONS}/${id}/cancel`)
    return response.data
  },
}

File: ms-reservation/src/main/java/com/example/ReservationEvent/services/ReservationService.java
package com.example.ReservationEvent.services;

import com.example.ReservationEvent.clients.EventClient;
import com.example.ReservationEvent.messaging.ReservationEventPublisher;
import com.example.ReservationEvent.models.*;
import com.example.ReservationEvent.repository.ReservationRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ReservationService {

    private final ReservationRepository reservationRepository;
    private final EventClient eventClient;
    private final ReservationEventPublisher eventPublisher;

    /**
     * Récupère TOUTES les réservations de la base (Postgres)
     */
    public List<ReservationResponse> getAllReservations() {
        return reservationRepository.findAll().stream()
                .map(ReservationResponse::fromEntity)
                .toList();
    }

    @Transactional
    public ReservationResponse createReservation(CreateReservationRequest request) {
        // 1. Vérifier l'événement via ms-event
        Event event = eventClient.detail(request.getEventId());

        if (event == null) {
            throw new EntityNotFoundException("Événement introuvable dans le microservice ms-event");
        }

        if (event.getAvailablePlaces() != null && event.getAvailablePlaces() <= 0) {
            throw new IllegalStateException("Plus de places disponibles");
        }

        // 2. Créer la réservation
        Reservation reservation = Reservation.builder()
                .userId(request.getUserId())
                .eventId(request.getEventId())
                .status(ReservationStatus.PENDING)
                .seatNumber(request.getSeatNumber())
                .build();

        Reservation saved = reservationRepository.save(reservation);

        // 3. Publier l'événement RESERVATION_CREATED (fire-and-forget).
        // If RabbitMQ is momentarily down, the reservation is still persisted.
        eventPublisher.publishReservationCreated(saved);

        return ReservationResponse.fromEntity(saved);
    }

    public List<ReservationResponse> getReservationsByUser(String userId) {
        return reservationRepository.findByUserId(userId).stream()
                .map(ReservationResponse::fromEntity)
                .toList();
    }

    public List<ReservationResponse> getReservationsByEvent(Long eventId) {
        return reservationRepository.findByEventId(eventId).stream()
                .map(ReservationResponse::fromEntity)
                .toList();
    }

    public ReservationResponse getReservationById(Long id) {
        return reservationRepository.findById(id)
                .map(ReservationResponse::fromEntity)
                .orElseThrow(() -> new EntityNotFoundException("Réservation #" + id + " introuvable"));
    }

    @Transactional
    public ReservationResponse confirmReservation(Long id) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Réservation introuvable"));

        reservation.setStatus(ReservationStatus.CONFIRMED);
        return ReservationResponse.fromEntity(reservationRepository.save(reservation));
    }

    @Transactional
    public ReservationResponse cancelReservation(Long id, User currentUser) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Réservation introuvable"));

        reservation.setStatus(ReservationStatus.CANCELLED);
        reservation.setCancelledAt(LocalDateTime.now());

        return ReservationResponse.fromEntity(reservationRepository.save(reservation));
    }

}